IPython/extensions/deduperreload/deduperreload.py

```diff
@@ -0,0 +1,477 @@
+                classes[ast_elt.name] = ast_elt
+            elif isinstance(ast_elt, ast.If):
+                unfixable.append(ast_elt.test)
+                if_defs, if_classes, if_unfixable = cls._gather_children(ast_elt.body)
+                else_defs, else_classes, else_unfixable = cls._gather_children(
+                    ast_elt.orelse
+                )
+                defs.update(if_defs)
+                defs.update(else_defs)
+                classes.update(if_classes)
+                classes.update(else_classes)
+                unfixable.extend(if_unfixable)
+                unfixable.extend(else_unfixable)
+            elif isinstance(ast_elt, (ast.AsyncWith, ast.With)):
+                unfixable.extend(ast_elt.items)
+                with_defs, with_classes, with_unfixable = cls._gather_children(
+                    ast_elt.body
+                )
+                defs.update(with_defs)
+                classes.update(with_classes)
+                unfixable.extend(with_unfixable)
+            elif isinstance(ast_elt, ast.Try):
+                try_defs, try_classes, try_unfixable = cls._gather_children(
+                    ast_elt.body
+                )
+                else_defs, else_classes, else_unfixable = cls._gather_children(
+                    ast_elt.orelse
+                )
+                finally_defs, finally_classes, finally_unfixable = cls._gather_children(
+                    ast_elt.finalbody
+                )
+                defs.update(try_defs)
+                defs.update(else_defs)
+                defs.update(finally_defs)
+                classes.update(try_classes)
+                classes.update(else_classes)
+                classes.update(finally_classes)
+                unfixable.extend(try_unfixable)
+                unfixable.extend(else_unfixable)
+                unfixable.extend(finally_unfixable)
+                for handler in ast_elt.handlers:
+                    if handler.type is not None:
+                        unfixable.append(handler.type)
+                    (
+                        handler_defs,
+                        handler_classes,
+                        handler_unfixable,
+                    ) = cls._gather_children(handler.body)
+                    defs.update(handler_defs)
+                    classes.update(handler_classes)
+                    unfixable.extend(handler_unfixable)
+            elif not isinstance(ast_elt, (ast.Ellipsis, ast.Pass)):
+                unfixable.append(ast_elt)
+        return defs, classes, unfixable
+
+    def detect_autoreload(
+        self,
+        old_node: ast.Module | ast.ClassDef,
+        new_node: ast.Module | ast.ClassDef,
+        prefixes: list[str] | None = None,
+    ) -> bool:
+        """
+        Returns `true` if we can run our targeted autoreload algorithm safely.
+        Returns `false` if we should instead use IPython's original autoreload implementation.
```

Comment:
```suggestion
        Returns
        -------
        `True` if we can run our targeted autoreload algorithm safely.
        `False` if we should instead use IPython's original autoreload implementation.
```

We try to follow [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) standard when we can.